crates/sui-bridge/src/sui_transaction_builder.rs

```diff
@@ -0,0 +1,225 @@
+    let type_tags = TYPE_TAGS.get_or_init(|| {
+        let package_id = get_bridge_package_id();
+        // println!("{:?}", format!("{:?}::btc::BTC", package_id));
+        let mut type_tags = HashMap::new();
+        type_tags.insert(TokenId::Sui, TypeTag::from_str("0x2::sui::SUI").unwrap());
+        type_tags.insert(
+            TokenId::BTC,
+            TypeTag::from_str(&format!("{:?}::btc::BTC", package_id)).unwrap(),
+        );
+        type_tags.insert(
+            TokenId::ETH,
+            TypeTag::from_str(&format!("{:?}::eth::ETH", package_id)).unwrap(),
+        );
+        type_tags.insert(
+            TokenId::USDC,
+            TypeTag::from_str(&format!("{:?}::usdc::USDC", package_id)).unwrap(),
+        );
+        type_tags.insert(
+            TokenId::USDT,
+            TypeTag::from_str(&format!("{:?}::usdt::USDT", package_id)).unwrap(),
+        );
+        type_tags
+    });
+    type_tags.get(&token_id).unwrap().clone()
+}
+
+pub fn build_transaction(
+    client_address: SuiAddress,
+    gas_object_ref: &ObjectRef,
+    action: VerifiedCertifiedBridgeAction,
+) -> BridgeResult<TransactionData> {
+    match action.data() {
+        BridgeAction::EthToSuiBridgeAction(_) | BridgeAction::SuiToEthBridgeAction(_) => {
+            build_token_bridge_approve_transaction(client_address, gas_object_ref, action)
+        }
+    }
+}
+
+fn build_token_bridge_approve_transaction(
+    client_address: SuiAddress,
+    gas_object_ref: &ObjectRef,
+    action: VerifiedCertifiedBridgeAction,
+) -> BridgeResult<TransactionData> {
+    let (bridge_action, sigs) = action.into_inner().into_data_and_sig();
+    let mut builder = ProgrammableTransactionBuilder::new();
+
+    let (source_chain, seq_num, sender, target_chain, target, token_type, amount) =
+        match bridge_action {
+            BridgeAction::SuiToEthBridgeAction(a) => {
+                let bridge_event = a.sui_bridge_event;
+                (
+                    bridge_event.sui_chain_id,
+                    bridge_event.nonce,
+                    bridge_event.sui_address.to_vec(),
+                    bridge_event.eth_chain_id,
+                    bridge_event.eth_address.to_fixed_bytes().to_vec(),
+                    bridge_event.token_id,
+                    bridge_event.amount,
+                )
+            }
+            BridgeAction::EthToSuiBridgeAction(a) => {
+                let bridge_event = a.eth_bridge_event;
+                (
+                    bridge_event.eth_chain_id,
+                    bridge_event.nonce,
+                    bridge_event.eth_address.to_fixed_bytes().to_vec(),
+                    bridge_event.sui_chain_id,
+                    bridge_event.sui_address.to_vec(),
+                    bridge_event.token_id,
+                    bridge_event.amount,
+                )
+            }
+        };
+
+    let source_chain = builder.pure(source_chain as u8).map_err(|e| {
```

Comment:
Why isn't the chain id already u8?
Similar question for the other casts

Reply:
Also, serializing numbers should really never fail

Reply:
they are enums to be less error prone, and with `#[repr(u8)]` for casting ability to u8. e.g. 

```

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone, Copy, TryFromPrimitive)]
#[repr(u8)]
pub enum BridgeChainId {
    SuiMainnet = 0,
    SuiTestnet = 1,
    SuiDevnet = 2,

    EthMainnet = 10,
    EthSepolia = 11,
}

```